tsdb/tsm1/reader_index.go

```diff
@@ -521,25 +533,94 @@ func (d *indirectIndex) DeletePrefix(prefix []byte, minTime, maxTime int64, dead
 
 		// Otherwise, we have to track it in the prefix tombstones list.
 		mustTrack = true
+
+		// If we have a predicate, we must keep track of a pending tombstone entry for the key.
+		if pred != nil {
+			pending = append(pending, pendingTombstone{
+				Key:         len(keys),
+				Index:       iter.Index(),
+				Offset:      offset,
+				EntryOffset: entryOffset,
+				Tombstones:  len(d.tombstones[offset]) + d.prefixTombstones.Count(key),
+			})
+			keys = append(keys, key)
+		}
 	}
 	d.mu.RUnlock()
 
 	// Check and abort if nothing needs to be done.
-	if !mustTrack && !iter.HasDeletes() {
+	if !mustTrack && len(pending) == 0 && !iter.HasDeletes() {
 		return false
 	}
 
 	d.mu.Lock()
 	defer d.mu.Unlock()
 
-	if mustTrack {
-		d.prefixTombstones.Append(prefix, TimeRange{Min: minTime, Max: maxTime})
+	if pred == nil {
+		// If we don't have a predicate, we can add a single prefix tombstone entry.
+		if mustTrack {
+			d.prefixTombstones.Append(prefix, TimeRange{Min: minTime, Max: maxTime})
+		}
+
+		// Clean up any fully deleted keys.
+		if iter.HasDeletes() {
+			iter.Done()
+		}
+		return true
+	}
+
+	// Otherwise, we must walk the pending deletes individually.
+	for _, p := range pending {
+		key := keys[p.Key]
+
+		// Check the existing tombstones. If the length did not change, then we know
+		// that we don't need to double check coverage, since we only ever increase the
+		// number of tombstones for a key.
+		if trs := d.tombstones[p.Offset]; p.Tombstones == len(trs)+d.prefixTombstones.Count(key) {
+			d.tombstones[p.Offset] = insertTimeRange(trs, minTime, maxTime)
+			continue
+		}
+
+		// Since the length changed, we have to do the expensive overlap check again.
+		// We re-read the entries again under the write lock because this should be
+		// rare and only during concurrent deletes to the same key. We could make
+		// a copy of the entries before getting here, but that penalizes the common
+		// no-concurrent case.
+		entries, err = readEntriesTimes(d.b.access(p.EntryOffset, 0), entries)
+		if err != nil {
```

Comment:
Could we at least log this?

Reply:
I'll attempt to introduce loggers into the TSMReader. That's going to be quite a bit of plumbing, unfortunately.

Why don't we just use package global loggers, again? :smile: